### PR TITLE
Implementation of the NVPT for APTs and AATs in velocity form

### PR DESCRIPTION
**Pull request by edditler**
Dear CP2K Developers,

This pull-request includes the implementation of the Nuclear Velocity Perturbation Theory (NVPT) as described in [doi.org/10.1021/acs.jctc.2c00006](doi.org/10.1021/acs.jctc.2c00006). The perturbed MO coefficients (with respect to nuclear velocities) are used to calculate the Atomic Polar/Axial Tensors in the velocity representation. The resulting output can be used, together with a Hessian matrix, to calculate vibrational circular dichroism spectra.

The previous implementation of the nuclear displacement perturbation is contained in the new `vcd_env` as `vcd_env%dcdr_env`.

### CODE REVIEW

File: src/aobasis/ai_moments.F [hunk 1]
```diff
@@ -46,10 +46,89 @@ MODULE ai_moments
...
+      INTEGER, INTENT(IN)                                :: lb_min, order
+      REAL(KIND=dp), DIMENSION(3), INTENT(IN)            :: rac, rbc
+      REAL(KIND=dp), DIMENSION(:, :, :, :), INTENT(OUT)  :: difmab
+      INTEGER, INTENT(IN)                                :: lambda
+      INTEGER, INTENT(IN), OPTIONAL                      :: iatom, jatom
+
+      INTEGER                                            :: alpha, beta, lda, lda_min, ldb, ldb_min
+      REAL(KIND=dp)                                      :: dab, rab(3)
+      REAL(KIND=dp), ALLOCATABLE, DIMENSION(:, :, :)     :: difmab_tmp
+      REAL(KIND=dp), DIMENSION(:, :, :), POINTER         :: mab
```
Comment (fstein93) [inline src/aobasis/ai_moments.F]: You can use an ALLOCATABLE instead of a POINTER, here.

File: src/commutator_rpnl.F [hunk 2]
```diff
@@ -434,22 +439,27 @@ SUBROUTINE build_com_mom_nl(qs_kind_set, sab_all, sap_ppnl, eps_ppnl, particle_s
       TYPE(cell_type), INTENT(IN), POINTER               :: cell
       TYPE(dbcsr_p_type), DIMENSION(:), INTENT(INOUT), &
          OPTIONAL, POINTER                               :: matrix_rv, matrix_rxrv, matrix_rrv
+      TYPE(dbcsr_p_type), DIMENSION(:, :), &
+         INTENT(INOUT), OPTIONAL, POINTER                :: matrix_r_rxvr, matrix_rxvr_r, &
+                                                            matrix_r_doublecom
```
Comment (fstein93) [inline src/commutator_rpnl.F:448]: Do these arrays really have to be POINTERs?

File: src/commutator_rpnl.F [hunk 3]
```diff
@@ -460,30 +470,56 @@ SUBROUTINE build_com_mom_nl(qs_kind_set, sab_all, sap_ppnl, eps_ppnl, particle_s
...
+      i_z = 4
+      i_xx = 5
+      i_xy = 6
+      i_xz = 7
+      i_yx = i_xy
+      i_yy = 8
+      i_yz = 9
+      i_zx = i_xz
+      i_zy = i_yz
+      i_zz = 10
```
Comment (fstein93) [inline src/commutator_rpnl.F]: Make them PARAMETERs

File: src/commutator_rpnl.F [hunk 4]
```diff
@@ -1167,4 +1615,241 @@ SUBROUTINE build_com_nl_mag(qs_kind_set, sab_all, sap_ppnl, eps_ppnl, particle_s
...
+
+!$    INTEGER(kind=omp_lock_kind), &
+!$       ALLOCATABLE, DIMENSION(:) :: locks
+!$    INTEGER                                            :: lock_num, hash
+!$    INTEGER, PARAMETER                                 :: nlock = 501
+
+      i_1 = 1
+      i_x = 2
+      i_y = 3
+      i_z = 4
```
Comment (fstein93) [inline src/commutator_rpnl.F]: Make them PARAMETERs

File: src/commutator_rpnl.F [hunk 5]
```diff
@@ -1167,4 +1615,241 @@ SUBROUTINE build_com_nl_mag(qs_kind_set, sab_all, sap_ppnl, eps_ppnl, particle_s
...
+
+                  IF (ALL(cell_b + alist_bc%clist(kbc)%cell - alist_ac%clist(kac)%cell == 0)) THEN
+                     IF (alist_ac%clist(kac)%maxac*alist_bc%clist(kbc)%maxach < eps_ppnl) CYCLE
+                     acint => alist_ac%clist(kac)%acint
+                     bcint => alist_bc%clist(kbc)%acint
+                     achint => alist_ac%clist(kac)%achint
+                     bchint => alist_bc%clist(kbc)%achint
+                     na = SIZE(acint, 1)
+                     np = SIZE(acint, 2)
+                     nb = SIZE(bcint, 1)
```
Comment (fstein93) [inline src/commutator_rpnl.F]: You can use an ASSOCIATE construct here

File: src/qs_dcdr_utils.F [hunk 6]
```diff
@@ -977,8 +982,8 @@ SUBROUTINE dcdr_env_init(dcdr_env, qs_env)
       CALL cp_print_key_finished_output(output_unit, logger, lr_section, &
                                         "PRINT%PROGRAM_RUN_INFO")
 
-      IF (dcdr_env%output_unit > 0) &
-         WRITE (dcdr_env%output_unit, *) 'List of atoms: ', dcdr_env%list_of_atoms
+      ! IF (dcdr_env%output_unit > 0) &
+      !    WRITE (dcdr_env%output_unit, *) 'List of atoms: ', dcdr_env%list_of_atoms
```
Comment (fstein93) [inline src/qs_dcdr_utils.F]: Do you still need it? If not, then remove it.

File: src/qs_mfp.F [hunk 7]
```diff
@@ -0,0 +1,1416 @@
...
+
+      INTEGER                                            :: alpha, delta, gamma, handle, ikind, &
+                                                            ispin, nao, nmo
+      LOGICAL                                            :: ghost
+      REAL(dp)                                           :: aat_linmom, aat_moment, aat_moment_der, &
+                                                            aat_overlap, aat_tmp, charge, f_spin, &
+                                                            lc_tmp, tmp_trace
+      REAL(dp), DIMENSION(3)                             :: gauge_origin
+      REAL(dp), DIMENSION(:, :, :), POINTER              :: aat_atom
+      TYPE(cp_fm_type), POINTER                          :: buf, buf2, matrix_dSdB_mo, mo_coeff
```
Comment (fstein93) [inline src/qs_mfp.F]: Maybe, you can drop this POINTER attribute.

File: src/qs_mfp.F [hunk 8]
```diff
@@ -0,0 +1,1416 @@
...
+      TYPE(neighbor_list_set_p_type), DIMENSION(:), &
+         POINTER                                         :: sab_all
+      TYPE(particle_type), DIMENSION(:), POINTER         :: particle_set
+      TYPE(qs_kind_type), DIMENSION(:), POINTER          :: qs_kind_set
+
+      CALL timeset(routineN, handle)
+
+      ! Some setup
+      ispin = 1
+      f_spin = 2._dp
```
Comment (fstein93) [inline src/qs_mfp.F]: These could be PARAMETERs.

File: src/qs_mfp.F [hunk 9]
```diff
@@ -0,0 +1,1416 @@
...
+      nmo = vcd_env%dcdr_env%nmo(1)
+      nao = vcd_env%dcdr_env%nao
+      mo_coeff => vcd_env%dcdr_env%mo_coeff(1)
+      aat_atom => vcd_env%aat_atom_mfp
```
Comment (fstein93) [inline src/qs_mfp.F]: Another potential case for ASSOCIATE.

File: src/qs_mfp.F [hunk 10]
```diff
@@ -0,0 +1,1416 @@
...
+      TYPE(qs_environment_type), POINTER                 :: qs_env
+      INTEGER                                            :: alpha
+
+      CHARACTER(LEN=*), PARAMETER :: routineN = 'mfp_build_operator_gauge_dependent'
+
+      INTEGER                                            :: delta, gamma, handle, ispin, nao, nmo
+      REAL(dp)                                           :: eps_ppnl, lc_tmp
+      REAL(dp), DIMENSION(3)                             :: gauge_origin
+      TYPE(cell_type), POINTER                           :: cell
+      TYPE(cp_fm_type), POINTER                          :: buf, mo_coeff
```
Comment (fstein93) [inline src/qs_mfp.F]: Drop the POINTER

**Review by fstein93 — COMMENTED** [review]
I made some comments. I did not mark everything. I especially ask you to give parameters the PARAMETER attribute and check whether you can drop some POINTER attributes or replace them with ALLOCATABLE attributes.

**Comment by edditler** [comment]
Great, thanks a lot!

File: src/qs_moments.F [hunk 11]
```diff
@@ -212,98 +205,101 @@ SUBROUTINE dipole_velocity_deriv(qs_env, difdip, order, lambda, rc)
...
-         DO j = 1, 3
-            NULLIFY (mint(i, j)%block)
-            CALL dbcsr_get_block_p(matrix=difdip(i, j)%matrix, &
-                                   row=irow, col=icol, BLOCK=mint(i, j)%block, &
-                                   found=found)
-            CPASSERT(found)
-            mint(i, j)%block = 0._dp
-         END DO
-         END DO
+         ASSOCIATE ( &
```
Comment (edditler) [inline src/qs_moments.F:207]: Should the ASSOCIATE construct also be used here?

Comment (fstein93) [reply src/qs_moments.F:207]: I think it does make sense here. I am aware that your former suggestion corresponds to the legacy coding style in CP2K. But with the ASSOCIATE construct it is clear that the variables are runtime constants.

Comment (edditler) [reply src/qs_moments.F:207]: Okay, I replaced the pointer assignments in all files that are touched by this PR.

**Comment by edditler** [comment]
It appears that either fpretty has an issue with aligning `ASSOCIATED` constructs in some cases, or I made a mistake.

Is there anything else left to do?

**Comment by fstein93** [comment]
Usually, it works quite well because `ASSOCIATE` should in that respect behave like `SUBROUTINE`, `FOR` or `IF`.

If the tests run smoothly, I will merge your PR.

**Comment by glb96** [comment]
Hi! 
Since this last PR, I have had some trouble printing moments. Please see the input file and the error below. 

Maybe the input parameters shall be changed after your update? 

Best,
Guillaume

#################
Input file:
&FORCE_EVAL
  METHOD QUICKSTEP
  &DFT
    &REAL_TIME_PROPAGATION
       MAX_ITER 10
       MAT_EXP TAYLOR
       EPS_ITER 1.0E-9
       INITIAL_WFN SCF_WFN
    &END
    &EFIELD
       INTENSITY 1.0E+10
       PHASE 0.0
       POLARISATION 0 1 0
       !520ev = 2.38 in nm
       WAVELENGTH [nm] 2.3752
       ENVELOP GAUSSIAN
       &GAUSSIAN_ENV
         SIGMA [fs] 0.1
         T0 [fs] 0.5
       &END
    &END
    BASIS_SET_FILE_NAME BASIS_MOLOPT
    BASIS_SET_FILE_NAME BASIS_SET
    POTENTIAL_FILE_NAME POTENTIAL
    &MGRID
      CUTOFF 100
    &END MGRID
    &QS
      EPS_DEFAULT 1.0E-10
    &END QS
    &SCF
      MAX_SCF 250
      EPS_SCF 1.0E-8
      SCF_GUESS ATOMIC
      &OUTER_SCF
      EPS_SCF 1.0E-3
      &END
      &OT ON
        ROTATION
      &END OT
    &END SCF
    &POISSON
      PERIODIC NONE
      POISSON_SOLVER MT
    &END POISSON
    &XC
      &XC_FUNCTIONAL PADE
      &END XC_FUNCTIONAL
    &END XC
    &PRINT
      &MOMENTS
         PERIODIC .FALSE.
         COMMON_ITERATION_LEVELS 100000
         MAX_MOMENT 2
         COM_NL .TRUE.
         VEL_REPRS .TRUE.
         &EACH
            MD 1
         &END EACH
      &END MOMENTS
    &END PRINT
  &END DFT
  &SUBSYS
    &CELL
      ABC 7 7 7
      PERIODIC NONE
    &END CELL
    &TOPOLOGY
      &CENTER_COORDINATES
      &END CENTER_COORDINATES
    &END TOPOLOGY
    &COORD
    O   0.000000    0.000000   -0.065587 H2O
    H   0.000000   -0.757136    0.520545 H2O
    H   0.000000    0.757136    0.520545 H2O
    &END COORD
    &KIND H
      BASIS_SET DZVP-MOLOPT-GTH
      POTENTIAL GTH-PADE-q1
    &END KIND
    &KIND O
      BASIS_SET DZVP-MOLOPT-GTH
      POTENTIAL GTH-PADE-q6
    &END KIND
  &END SUBSYS
&END FORCE_EVAL
&GLOBAL
  PROJECT H2O-rtp-efield
 RUN_TYPE RT_PROPAGATION
  PRINT_LEVEL LOW
&END GLOBAL
&MOTION
  &MD
    ENSEMBLE NVE
    STEPS 10
    TIMESTEP [fs] 0.0005
    TEMPERATURE 300.0
  &END MD
    &PRINT
       &TRAJECTORY
          &EACH
             MD 1
          &END
       &END
    &END
&END MOTION
###########################################
Error: 
At line 946 of file /users/glebreton/Software/real_cp2k/src/qs_moments.F
Fortran runtime error: Index '4' of dimension 1 of array 'moments_der' above upper bound of 3

Error termination. Backtrace:
#0  0x7fb8fc0e5640 in ???
#1  0x7fb8fc0e6185 in ???
#2  0x7fb8fc0e652a in ???
#3  0xe6e520 in __qs_moments_MOD_build_local_moments_der_matrix
        at /users/glebreton/Software/real_cp2k/src/qs_moments.F:946
#4  0xe77cfa in __qs_moments_MOD_qs_moment_locop
        at /users/glebreton/Software/real_cp2k/src/qs_moments.F:2290
#5  0xf67d49 in __qs_scf_post_gpw_MOD_qs_scf_post_moments
        at /users/glebreton/Software/real_cp2k/src/qs_scf_post_gpw.F:1257
#6  0xf6a595 in __qs_scf_post_gpw_MOD_scf_post_calculation_gpw
        at /users/glebreton/Software/real_cp2k/src/qs_scf_post_gpw.F:377
#7  0xf7477d in __qs_scf_post_scf_MOD_qs_scf_compute_properties
        at /users/glebreton/Software/real_cp2k/src/qs_scf_post_scf.F:78
#8  0xf32951 in __qs_scf_MOD_scf
        at /users/glebreton/Software/real_cp2k/src/qs_scf.F:300
#9  0xce0d43 in __qs_energy_MOD_qs_energies
        at /users/glebreton/Software/real_cp2k/src/qs_energy.F:111
#10  0xd11ae6 in __qs_force_MOD_qs_calc_energy_force
        at /users/glebreton/Software/real_cp2k/src/qs_force.F:112
#11  0x9b5e83 in __force_env_methods_MOD_force_env_calc_energy_force
        at /users/glebreton/Software/real_cp2k/src/force_env_methods.F:257
#12  0x5d34d2 in rt_initial_guess
        at /users/glebreton/Software/real_cp2k/src/motion/rt_propagation.F:457
#13  0x5d45d9 in __rt_propagation_MOD_rt_prop_setup
        at /users/glebreton/Software/real_cp2k/src/motion/rt_propagation.F:134
#14  0x45620f in cp2k_run
        at /users/glebreton/Software/real_cp2k/src/start/cp2k_runs.F:382
#15  0x4578cf in __cp2k_runs_MOD_run_input
        at /users/glebreton/Software/real_cp2k/src/start/cp2k_runs.F:996
#16  0x4523c3 in cp2k
        at /users/glebreton/Software/real_cp2k/src/start/cp2k.F:379
#17  0x45252f in main
        at /users/glebreton/Software/real_cp2k/src/start/cp2k.F:44
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpirun detected that one or more processes exited with non-zero status, thus causing
the job to be terminated. The first process to do so was:

  Process name: [[59889,1],0]
  Exit code:    2
--------------------------------------------------------------------------

**Comment by edditler** [comment]
The error is probably due to this change: https://github.com/cp2k/cp2k/pull/2568/files#diff-878e9543a198580fa55ff8b5adf912a7685b7c588d6fccbfeb396e0c8378391aL408 (in master https://github.com/cp2k/cp2k/blob/306f83e90954f601340e42b0b38991b65a4f1f51/src/qs_moments.F#L840)

My understanding is, that `nmoments` and `nder`ivatives were switched up previously (logically; the code was obviously working).

In https://github.com/cp2k/cp2k/blob/master/src/qs_moments.F#L2280-L2290 it looks like the dipole moments and their first derivatives are allocated, but the quadrupoles and their first derivatives are requested from `build_local_moments_der_matrix`.

**Comment by oschuett** [comment]
This PR has also introduced a number of [coding convention problems](https://dashboard.cp2k.org/archive/conventions/commit_f387d9e52db66ef3f179c70ace9df802df4c9cad.txt).


For example, `nspin` is not initialized in [qs_mfp.F](https://github.com/cp2k/cp2k/blob/master/src/qs_mfp.F) and there is a rough `PRINT` statement in https://github.com/cp2k/cp2k/blob/f387d9e52db66ef3f179c70ace9df802df4c9cad/src/qs_vcd_utils.F#L631

**Comment by edditler** [comment]
I will have a look at those. Sorry about that!

**Pull request merged**